cranelift/codegen/src/isa/x64/encoding/evex.rs

```diff
@@ -151,10 +178,44 @@ impl EvexInstruction {
     /// - finally, the ModR/M byte.
     ///
     /// Eventually this method should support encodings of more than just the reg-reg addressing mode (TODO).
-    pub fn encode<CS: ByteSink + ?Sized>(&self, sink: &mut CS) {
+    pub fn encode(&self, sink: &mut MachBuffer<Inst>) {
+        if let RegisterOrAmode::Amode(amode) = &self.rm {
+            if amode.can_trap() {
+                sink.add_trap(TrapCode::HeapOutOfBounds);
+            }
+        }
         sink.put4(self.bits);
         sink.put1(self.opcode);
-        sink.put1(encode_modrm(3, self.reg.0 & 7, self.rm.0 & 7));
+
+        match &self.rm {
+            RegisterOrAmode::Register(reg) => {
+                let rm: u8 = (*reg).into();
+                sink.put1(rex::encode_modrm(3, self.reg.0 & 7, rm & 7));
+            }
+            RegisterOrAmode::Amode(amode) => {
+                // NB: the `scaling` factor must be passed to
+                // `emit_modrm_sib_disp` because EVEX instructions always use a
+                // scaling factor. Section 2.7.5 in the Intel manual has a lot
+                // of words about this I don't fully understand yet. That being
+                // said for the V128 case the factor is 16 (or so I'm pretty
+                // sure). Assert that that's the "length" of this instruction
+                // (aka the encoding of the default `EvexContext`). If
+                // this assertion trips then the scaling factor probably needs
+                // to be stored on `self` and reassigned whenever the `length`
+                // method is called.
+                let b128 = EvexContext::Other {
+                    length: EvexVectorLength::V128,
+                };
+                let bits = (self.bits >> Self::LL.start()) & ((1 << Self::LL.len()) - 1);
```

Comment:
I would probably create a private `read` function for this just to avoid confusion.